### PR TITLE
big gRPC change WIP

### PR DESCRIPTION
**Pull request by drgarcia1986**
still in `WIP`

---
This change is [Reviewable](https://reviewable.io/reviews/luizalabs/teresa-api/182)

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+		return nil, err
+	}
+
+	s := grpc.NewServer(grpc.Creds(creds))
+
+	cs := user.NewServer(user.NewDatabaseOperations(opt.DB, opt.RSAPrivateKey))
+	cs.RegisterServer(s)
+
+	return &Server{listener: l, grpcServer: s}, nil
+}
```
Comment (aguerra) [inline pkg/server/server.go]: Maybe use envconfig here (or in main) to configure TLSCert, TLSKey

File: pkg/client/client.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+func New(cfg ClusterConfig) (*grpc.ClientConn, error) {
+	tlsConfig := &tls.Config{InsecureSkipVerify: true}
+	creds := credentials.NewTLS(tlsConfig)
+	return grpc.Dial(
+		cfg.Server,
+		grpc.WithPerRPCCredentials(&tokenAuth{cfg.Token}),
+		grpc.WithTransportCredentials(creds),
+	)
+}
```
Comment (aguerra) [inline pkg/client/client.go]: Are `GetRequestMetadata` `RequireTransportSecurity` grpc stuff?

File: cmd/client/cmd/login_cluster.go [hunk 3]
```diff
@@ -2,12 +2,20 @@ package cmd
...
+	context "golang.org/x/net/context"
 
-	"github.com/go-openapi/strfmt"
 	"github.com/howeyc/gopass"
+	"github.com/luizalabs/teresa-api/cmd/client/connection"
+	"github.com/luizalabs/teresa-api/pkg/client"
 	"github.com/spf13/cobra"
+
+	userpb "github.com/luizalabs/teresa-api/pkg/protobuf"
 )
```
Comment (aguerra) [inline cmd/client/cmd/login_cluster.go]: Forget it, userpb is ok.

**Comment by drgarcia1986** [comment]
DON'T MERGE!

File: pkg/client/client.go (hunk 2 above)
Comment (drgarcia1986) [inline pkg/client/client.go]: Yep, from the [PerRPCCredentials](https://godoc.org/google.golang.org/grpc/credentials#PerRPCCredentials) interface.

File: pkg/server/server.go (hunk 1 above)
Comment (drgarcia1986) [inline pkg/server/server.go]: `main.go` it's the right answer (I think), but, I don't know how we'll deal with TLSCerts (on secrets, I know, but, in files or bytes?)

File: pkg/server/user/handlers.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	return &userpb.LoginResponse{Token: token}, nil
+}
+
+func (s *Server) RegisterServer(grpcServer *grpc.Server) {
+	userpb.RegisterUserServer(grpcServer, s)
+}
+
+func NewServer(ops Operations) *Server {
+	return &Server{ops: ops}
+}
```
Comment (drgarcia1986) [inline pkg/server/user/handlers.go]: Agree

**Comment by drgarcia1986** [comment]
Agree with "pkg common" comment, but, are you have any suggestion for that case?

File: cmd/client/cmd/login_cluster.go (hunk 3 above)
Comment (drgarcia1986) [inline cmd/client/cmd/login_cluster.go]: You're right `protobuf` is the common package for all protobuf, but, I just don't like `pb` as package name on big code bases 😂

Comment (aguerra) [inline cmd/client/cmd/login_cluster.go]: :+1:

File: pkg/server/server.go (hunk 1 above)
Comment (aguerra) [inline pkg/server/server.go]: I prefer something like `export TLS_CERT=/etc/server.crt`

File: pkg/client/client.go (hunk 2 above)
Comment (aguerra) [inline pkg/client/client.go]: :+1:

File: pkg/server/user/handlers.go (hunk 4 above)
Comment (drgarcia1986) [inline pkg/server/user/handlers.go]: `:GoRename` ❤️

**Comment by aguerra** [comment]
For now I would move `pkg/common/errors.go -> pkg/errors/errors.go`, but you are right, it will not be always possible to avoid common.

**Comment by drgarcia1986** [comment]
If we do this, we need to deal with conflicts with standard library (`errors` package)

File: pkg/server/server.go (hunk 1 above)
Comment (drgarcia1986) [inline pkg/server/server.go]: LGTM, but are you thinking to change RSA key schema? 
I mean, we need to deal in the same way for both secrets types.

Comment (aguerra) [inline pkg/server/server.go]: I would mount secrets as files and only deal with files on code, but would postpone this change if I couldn't do it in all places (too big, too risky, etc).

**Comment by drgarcia1986** [comment]
@aguerra to avoid `common` pkg, what about move `ErrPermisionDenied` to `server/user` ? We're talking about users right?

**Pull request closed without merging**